crates/adapters/src/integrated/postgres/output.rs

```diff
@@ -243,14 +288,36 @@ impl PostgresOutputEndpoint {
         ))
     }
 
-    fn exec_statement(
+    fn exec_statement(&mut self, stmt: Statement, value: &mut Vec<u8>, name: &str) {
+        if let Err(e) = self.exec_statement_inner(stmt.clone(), value, name) {
+            let retry = e.should_retry();
+            let Some(controller) = self.controller.upgrade() else {
+                tracing::warn!("failed to upgrade the controller");
+                return;
+            };
+
+            controller.output_transport_error(
+                self.endpoint_id,
+                &self.endpoint_name,
+                true,
+                e.inner(),
+            );
+
+            if !retry {
+                return;
+            }
+
+            self.retry_connecting_with_backoff();
+            self.exec_statement(stmt, value, name)
```

Comment:
Rust doesn't guarantee that tail recursion gets compiled into a loop, so this needs to be written as an explicit loop.